http_proxy.go

```diff
@@ -500,79 +504,116 @@ func (hp *HTTPProxy) handler() http.Handler {
 }
 
 func (hp *HTTPProxy) Run(ctx context.Context) error {
-	var srv *http.Server
-
-	var wg sync.WaitGroup
-	wg.Add(1)
-
-	go func() {
-		defer wg.Done()
-
-		<-ctx.Done()
-		if srv != nil {
-			if err := srv.Shutdown(context.Background()); err != nil {
-				hp.log.Errorf("failed to shutdown server error=%s", err)
-			}
-		} else {
-			hp.Close()
-		}
-	}()
-
-	var srvErr error
 	if hp.config.TestingHTTPHandler {
 		hp.log.Infof("using http handler")
-		srv = &http.Server{
-			Handler:           hp.handler(),
-			IdleTimeout:       hp.config.IdleTimeout,
-			ReadTimeout:       hp.config.ReadTimeout,
-			ReadHeaderTimeout: hp.config.ReadHeaderTimeout,
-			WriteTimeout:      hp.config.WriteTimeout,
-		}
-		srvErr = srv.Serve(hp.listener)
-	} else {
-		srvErr = hp.proxy.Serve(hp.listener)
+		return hp.runHTTPHandler(ctx)
+	}
+	return hp.run(ctx)
+}
+
+func (hp *HTTPProxy) runHTTPHandler(ctx context.Context) error {
+	srv := http.Server{
+		Handler:           hp.handler(),
+		IdleTimeout:       hp.config.IdleTimeout,
+		ReadTimeout:       hp.config.ReadTimeout,
+		ReadHeaderTimeout: hp.config.ReadHeaderTimeout,
+		WriteTimeout:      hp.config.WriteTimeout,
 	}
-	if srvErr != nil {
-		if errors.Is(srvErr, net.ErrClosed) {
-			srvErr = nil
+
+	var g errgroup.Group
+	g.Go(func() error {
+		<-ctx.Done()
+		if err := srv.Shutdown(context.Background()); err != nil {
+			hp.log.Errorf("failed to shutdown server error=%s", err)
 		}
-		return srvErr
+		return ctx.Err()
+	})
+	for i := range hp.listeners {
+		l := hp.listeners[i]
+		g.Go(func() error {
+			return srv.Serve(l)
+		})
 	}
+	return g.Wait()
+}
 
-	wg.Wait()
-	return nil
+func (hp *HTTPProxy) run(ctx context.Context) error {
+	var g errgroup.Group
+	g.Go(func() error {
+		<-ctx.Done()
+		hp.Close()
+		return ctx.Err()
+	})
+	for i := range hp.listeners {
+		l := hp.listeners[i]
+		g.Go(func() error {
+			err := hp.proxy.Serve(l)
+			if errors.Is(err, net.ErrClosed) {
+				err = nil
+			}
+			return err
+		})
+	}
+	return g.Wait()
 }
 
-func (hp *HTTPProxy) listen() (net.Listener, error) {
+func (hp *HTTPProxy) listen() (_ []net.Listener, ferr error) {
 	switch hp.config.Protocol {
 	case HTTPScheme, HTTPSScheme, HTTP2Scheme:
 	default:
 		return nil, fmt.Errorf("invalid protocol %q", hp.config.Protocol)
 	}
 
-	l := Listener{
-		ListenerConfig: hp.config.ListenerConfig,
-		TLSConfig:      hp.tlsConfig,
-		PromConfig: PromConfig{
-			PromNamespace: hp.config.PromNamespace,
-			PromRegistry:  hp.config.PromRegistry,
-		},
+	prototype := Listener{
+		TLSConfig: hp.tlsConfig,
+		metrics:   newListenerMetrics(hp.config.PromRegistry, hp.config.PromNamespace),
```

Comment:
How about adding a `name` to the ListenerConfig and using it in metrics? That way we could have them more detailed e.g. which listener accepts the most.

Reply:
Sure it's a good idea.